### PR TITLE
Add non-mobile web support

### PR DESCRIPTION
**Pull request by ariabuckles**
Add basic support for mouseevents and keyboardevents.

We've been using these changes written originally by @allofthenorthwood and @spicyj in our build of perseus, but i'm porting them to the source files so i can stop working in the compiled perseus.js compiled file. I'm guessing they're probably good/useful for KA too.

CC @spicyj @crm416

### CODE REVIEW

File: src/components/input/math-input.js [hunk 1]
```diff
@@ -203,9 +203,9 @@ const MathInput = React.createClass({
         window.removeEventListener('touchstart', this.recordTouchStartOutside);
         window.removeEventListener('touchend', this.blurOnTouchEndOutside);
         window.removeEventListener('touchcancel', this.blurOnTouchEndOutside);
-        window.removeEventListener('resize', this._clearKeypadBoundsCache());
```
Comment (crm416) [inline src/components/input/math-input.js]: Ooh, thank you!

File: src/components/input/math-input.js [hunk 2]
```diff
@@ -255,36 +255,81 @@ const MathInput = React.createClass({
...
 
+    _postKeyEvent() {
+      // Trigger an `onChange` if the value in the input changed, and hide
+      // the cursor handle whenever the user types a key. If the value
+      // changed as a result of a keypress, we need to be careful not to
+      // call `setState` until after `onChange` has resolved.
+      const hideCursor = () => {
+          this.setState({
+              handle: {
+                  visible: false
```
Comment (crm416) [inline src/components/input/math-input.js]: Super nit: trailing commas in the objects please :)

File: src/components/input/math-input.js [hunk 3]
```diff
@@ -255,36 +255,81 @@ const MathInput = React.createClass({
         });
     },
 
+    _forwardGlobalKeydown(e) {
+        if (e.keyCode === 13 || e.keyCode === 27) { // Enter, Esc
+            this.props.keypadElement.dismiss();
```
Comment (crm416) [inline src/components/input/math-input.js]: Just to make sure I follow: by dismissing, we then blur, and thereby avoid forwarding the key-press event for an Enter or Escape down to MathQuill. Is that right?

File: src/components/input/math-input.js [hunk 4]
```diff
@@ -255,36 +255,81 @@ const MathInput = React.createClass({
...
+
+    _forwardGlobalKeypress(e) {
+        if (e.fakeForMathquill) {
+            return;
+        }
+        // Don't allow spaces, since they invalidate a lot of expressions in our grader
+        if (e.charCode === 32) {
+            return;
+        }
+        const e2 = new KeyboardEvent(e.type, e);
```
Comment (crm416) [inline src/components/input/math-input.js]: Thoughts on pulling this out into a helper fn?

File: src/fake-react-native-web/view.js [hunk 5]
```diff
@@ -74,6 +74,43 @@ const View = React.createClass({
...
+                    doc.removeEventListener('mouseup', onUp);
+                    augmentMouseEvent(e);
+                    this.props.onTouchEnd && this.props.onTouchEnd(e);
+                };
+                doc.addEventListener('mousemove', onMove, false);
+                doc.addEventListener('mouseup', onUp, false);
+
+                // Need to .persist() a React event object before adding
+                // properties to it since it's reused otherwise.
+                e.persist();
```
Comment (crm416) [inline src/fake-react-native-web/view.js]: Is this not necessary for the `onMove` and `onUp` cases?

File: src/components/input/math-input.js [hunk 6]
```diff
@@ -255,36 +255,81 @@ const MathInput = React.createClass({
...
+        Object.defineProperty(e2, 'keyCode', {get: () => e.keyCode});
+        Object.defineProperty(e2, 'charCode', {get: () => e.charCode});
+        Object.defineProperty(e2, 'which', {get: () => e.which});
+        e2.fakeForMathquill = true;
+        this.mathField.fakeTextarea.value = String.fromCharCode(e.charCode);
```
Comment (crm416) [inline src/components/input/math-input.js]: On initial read I'd expect this to continuously set the input to the value of the last-pressed key. Is that incorrect? :D

**Comment by moliata** [comment]
You should add "evt.preventDefault()" in onTouchEnd event at gesture-manager.js, so on mobile devices, key wouldn't be "pressed twice".

**Comment by nixterrimus** [comment]
@ariabuckles Thanks for this! We're just getting back to actively working on and maintaining this math input project. Lots has changed- we have new designs and goals. I'm going to close out your PR for now. Is there anything here you want to make sure we get into the production version?

**Comment by ariabuckles** [comment]
Ah, sorry @nixterrimus and @crm416 ! Nah, I sadly stopped using this on desktop web. If using this on desktop KA ever becomes a priority feel free to use parts of this as-is helpful, or rewrite if you like. If not no worries!

**Pull request closed without merging**